### PR TITLE
Fix Teeworlds server query, #3

### PR DESCRIPTION
**Pull request by illwieckz**
Hi, this fixes the Teeworlds server query, it tries with the 3 known `getinfo` packets.

I renamed the `-tee` argument to `-tees` since I will work next on the `-teem` Teeworlds master query.

Since Teeworlds support was broken since many years, there is probably nobody that uses the `-tee` arg today.

### CODE REVIEW

**Comment by illwieckz** [comment]
Hi, I've updated.

The part with the useless `strdup` was rewritten, and I added an extra test to verify that the whole packet had a NULL end in it. In fact, all `strnlen` called on packet parts are called with a safe maxlength, so this extra test is probably not necessary.

**Comment by stevenh** [comment]
Your strnlen tests where fine but add_rule uses strdup so if its not null terminated then "boom" ;-)

File: tee.c [hunk 1]
```diff
@@ -17,76 +18,229 @@
...
-		default:
-			add_rule( server, server->type->game_rule, "unknown", NO_FLAGS);
-			break;
+	/* packet too short */
+	if (len_teeserver_info_headerprefix > rawpktlen) {
+		return (PKT_ERROR);
+	}
+
+	/* not null-terminated packet */
+	if (strnlen(rawpkt, rawpktlen) == rawpktlen && rawpkt[rawpktlen - 1] != 0) {
```
Comment (stevenh) [inline tee.c]: Why the strlen, surely that will always fail as the last byte is expected to be 0 hence the length with be rawpktlen - 1?

File: tee.c [hunk 2]
```diff
@@ -17,76 +18,229 @@
...
+	/* get the last character */
+	last_char = rawpkt[len_teeserver_info_headerprefix];
+
+	/* compare the response without the last character, and verify if the last character is 'o', '2' or '3' */
+	if (memcmp(rawpkt + skip, teeserver_info_headerprefix, len_teeserver_info_headerprefix) != 0 || (last_char != 'o' && last_char != '2' && last_char != '3')) {
+		return (PKT_ERROR);
+	}
+
+	/* header, skip */
+	skip = len_teeserver_info_headerprefix + sizeof(last_char);
```
Comment (stevenh) [inline tee.c]: Any reason your not holding a current pointer instead of using rawpkt + skip all the time?

File: tee.c [hunk 3]
```diff
@@ -17,76 +18,229 @@
...
+		player->score = atoi(rawpkt + skip);
+		skip += strnlen(rawpkt + skip, rawpktlen - skip) + 1;
+		if (skip > rawpktlen) {
+			return (PKT_ERROR);
+		}
+	}
 
-		free(version);
+	/* version reprise */
+	server->protocol_version = 0;
```
Comment (stevenh) [inline tee.c]: Any reason why this isnt calculated as soon as you have the version pointer?

File: tee.c [hunk 4]
```diff
@@ -17,76 +18,229 @@
...
 	}
+	server->protocol_version |= (atoi(tok) & 0x000F) << 8;
+
+	tok = strtok(NULL, ".");
+	if (tok == NULL) {
+		return (PKT_ERROR);
+	}
+	server->protocol_version |= (atoi(tok) & 0x00FF);
+
+	free(version);
```
Comment (stevenh) [inline tee.c]: Leak here if you return early for any reason

File: tee.c (hunk 1 above)
Comment (illwieckz) [inline tee.c]: I wanted to test "there is no null before end" (length=maxlength) _and_ "last is not null"

Just to verify:

```c
$ cat a.c
#include <stdio.h>
#include <strings.h>

int main (void) {

	char tab1[3]={'a', 'b', 'c'};
	char tab2[3]={'a', 'b', 0};

	printf("not terminated: %d\n", strnlen(tab1, 3));
	printf("null terminated: %d\n", strnlen(tab2, 3));

}

$ gcc -o a a.c

$ ./a 
not terminated: 3
null terminated: 2
```

So, `strnlen` returns `rawpktlen` if not terminated string, first condition works and is sufficient, no need to test the last character.

So only:
```c
if (strnlen(rawpkt, rawpktlen) == rawpktlen) {
	return (PKT_ERROR);
}
```
is needed.

File: tee.c (hunk 2 above)
Comment (illwieckz) [inline tee.c]: Yes, I can probably increment `rawpkt` all the time and decrement `rawpktlen` all the time, I will try if it works everywhere, if I use `skip` as a temporary value to store the step to increment the pointer and the lenght with the same value each time there is an in/decrement to do.

File: tee.c (hunk 3 above)
Comment (illwieckz) [inline tee.c]: I don't know, it's one of the very last line from the original code. :stuck_out_tongue_winking_eye:

File: tee.c (hunk 4 above)
Comment (illwieckz) [inline tee.c]: These lines are the other lines I have not written (and probably the only lines remaining from the original code). I did not quite understand what they do, if you have a suggestion to prevent this leak, please suggest me. :wink:

File: tee.c (hunk 2 above)
Comment (illwieckz) [inline tee.c]: Instead of doing:

```c
		skip += strnlen(rawpkt + skip, rawpktlen - skip) + 1; 
		if (skip > rawpktlen) {
			return (PKT_ERROR);
		}
```

It seems I can do:
```
		skip = strnlen(rawpkt, rawpktlen) + 1; 
		rawpkt += skip;
		rawpktlen -= skip;
		if (skip > rawpktlen) {
			return (PKT_ERROR);
		}
```

I don't know if it is better.

Comment (illwieckz) [inline tee.c]: Since I check that `rawpkt` is null-terminated at the beginning, perhaps I can drop all the occurrences of this check:

```c
		if (skip > rawpktlen) {
			return (PKT_ERROR);
		}
```

File: tee.c (hunk 4 above)
Comment (illwieckz) [inline tee.c]: Hum, yes, I must free `version` before each `return`, but it seems there is no need to `strdup` at all. Right?

File: tee.c (hunk 2 above)
Comment (stevenh) [inline tee.c]: I would just setup a pointer pointing to the last byte at the start, something like:
```c
p = rawpkt
end = rawpkt + rawpktlen;
... stuff

p += strnlen(p, rawpktend - p) + 1; 
if (p > end) {
     malformed_packet(server, "short packet reading field XXXX");
     return (PKT_ERROR);
}
````

File: tee.c (hunk 4 above)
Comment (stevenh) [inline tee.c]: No not really, the data will be munged, as strtok does its work, the packet but your the only consumer so that doesn't matter.

File: tee.c (hunk 1 above)
Comment (stevenh) [inline tee.c]: No you still need the null check as strnlen will return rawpktlen even if it wasn't null terminated.

**Comment by stevenh** [comment]
Any update on this @illwieckz ?

**Comment by illwieckz** [comment]
Hi @stevenh, I do not forget this, it's just that I do not have much time these days. I'll work on it soon. :wink:

**Comment by stevenh** [comment]
No problem just checking :)

**Comment by illwieckz** [comment]
Just to say, I will probably be able to work on it in july or august. I’m not forgetting this, I’m just very busy. :-)

**Comment by illwieckz** [comment]
Hi @stevenh, I modified my patch, if you want to test it you can do that:

`qstat -default tees -f /tmp/teelist.txt`

With [this list](http://pastebin.com/WkAqY0CU) as `/tmp/teelist.txt`.

File: tee.c [hunk 5]
```diff
@@ -17,76 +18,187 @@
...
+	current = rawpkt;
+	end = rawpkt + rawpktlen;
+
+	/* header, skip */
+	current += len_teeserver_info_headerprefix + sizeof(last_char); 
+
+	/* if inf2 or inf3 */
+	if (last_char == '2' || last_char == '3') {
+		/* token, skip */
+		current += strnlen(current, end - current) + 1;
```
Comment (stevenh) [inline tee.c]: It looks like you could simply these tests by switching to maintaining a "remaining" instead of having to constantly check against end - current?

File: tee.c [hunk 6]
```diff
@@ -17,76 +18,187 @@
...
+		player->score = atoi(current);
+		current += strnlen(current, end - current) + 1;
+	}
+
+	/* version reprise */
+	server->protocol_version = 0;
+
+	tok = strtok(version, ".");
+	if (tok == NULL) {
+		return (PKT_ERROR);
```
Comment (stevenh) [inline tee.c]: Be nice to add info on what the pkt error is here and below so when run in verbose the issue is clear.

File: tee.c [hunk 7]
```diff
@@ -17,76 +18,187 @@
...
-	// skip unimplemented ack, crc, etc
-	char *pkt = rawpkt + 6;
-	char *tok = NULL, *version = NULL;
-	int i;
+	int i, skip;
+	char last_char;
+	char *current = NULL, *end = NULL, *version = NULL, *tok = NULL;
 	struct player* player;
 
 	server->ping_total += time_delta(&packet_recv_time, &server->packet_time1);
```
Comment (illwieckz) [inline tee.c]: woah, this line is the only line I did not rewrite and it's just boilerplate. :stuck_out_tongue:

File: tee.c [hunk 8]
```diff
@@ -17,76 +18,187 @@
...
+
+	/* if inf2 or inf3 */
+	if (last_char == '2' || last_char == '3') {
+		/* token, skip */
+		current += strnlen(current, end - current) + 1; 
+	}
+
+	/* version */
+	version = current;
+	current += strnlen(current, end - current) + 1;
```
Comment (illwieckz) [inline tee.c]: Do you think we need to do after each `strlen` and before the next token

```c
if (current == end) {
	return PKT_ERROR;
}
```

?

File: tee.c (hunk 5 above)
Comment (illwieckz) [inline tee.c]: Hmm, I don't know why, but my question disappeared, I was asking if I can just do that:

```c
current += strlen(current) + 1;
```

since I already checked that the packet is NULL-terminated.

File: tee.c (hunk 8 above)
Comment (stevenh) [inline tee.c]: Yep that would be a good idea

Comment (illwieckz) [inline tee.c]: OK, so, since the null-termination of the packet was verified and since we can check if the packet is ok each time we read a new value, can I use `strlen` and get rid of any `remaining` tracking?

Comment (stevenh) [inline tee.c]: Better to just compare current <= end as that avoids the strlen

Comment (illwieckz) [inline tee.c]: `remaining` needs the value returned by `strnlen` to be decremented.
`current` needs the value returned by `strnlen` to be decremented.

Unless I miss something, it seems there is only two way to do that:

* Using an `end` variable (what I do):

```c
current = rawpkt;
end = rawpkt + rawpktlen;

/* header, skip */
current += len_teeserver_info_headerprefix + sizeof(last_char);
if (current == end)
{
	return PKT_ERROR;
}

/* version */
version = current;
current += strnlen(current, end - current) + 1;
if (current == end)
{
	return PKT_ERROR;
}
```

* Using a `remaining` variable:

```c
current = rawpkt;
remaining = rawpktlen;

/* header, skip */
tokenlen = len_teeserver_info_headerprefix + sizeof(last_char);
current += tokenlen;
remaining -= tokenlen;
if (remaining == 0)
{
	return PKT_ERROR;
}

/* version */
version = current;
tokenlen += strnlen(current, remaining) + 1;
current += tokenlen;
remaining -= tokenlen;
if (remaining == 0)
{
	return PKT_ERROR;
}
```

But perhaps I miss something.

**Comment by skybon** [comment]
This PR is nearly 1 year old. Outline any remaining issues in PR's description perhaps?
